wgpu-core/src/instance.rs

```diff
@@ -360,71 +415,128 @@ impl<F: IdentityFilter<AdapterId>> Global<F> {
+                    raw: adapters.metal.0.swap_remove(selected),
                 };
                 log::info!("Adapter Metal {:?}", adapter.raw.info);
                 let id = backend::Metal::hub(self).adapters.register_identity(
-                    id_metal.unwrap(),
+                    adapters.metal.1.unwrap(),
                     adapter,
                     &mut token,
                 );
                 return Some(id);
             }
-            selected -= adapters_mtl.len();
+            selected -= adapters.metal.0.len();
         }
         #[cfg(windows)]
         {
-            if selected < adapters_dx12.len() {
+            if selected < adapters.dx12.0.len() {
                 let adapter = Adapter {
-                    raw: adapters_dx12.swap_remove(selected),
+                    raw: adapters.dx12.0.swap_remove(selected),
                 };
                 log::info!("Adapter Dx12 {:?}", adapter.raw.info);
                 let id = backend::Dx12::hub(self).adapters.register_identity(
-                    id_dx12.unwrap(),
+                    adapters.dx12.1.unwrap(),
                     adapter,
                     &mut token,
                 );
                 return Some(id);
             }
-            selected -= adapters_dx12.len();
-            if selected < adapters_dx11.len() {
+            selected -= adapters.dx12.0.len();
+            if selected < adapters.dx11.0.len() {
                 let adapter = Adapter {
-                    raw: adapters_dx11.swap_remove(selected),
+                    raw: adapters.dx11.0.swap_remove(selected),
                 };
                 log::info!("Adapter Dx11 {:?}", adapter.raw.info);
                 let id = backend::Dx11::hub(self).adapters.register_identity(
-                    id_dx11.unwrap(),
+                    adapters.dx11.1.unwrap(),
                     adapter,
                     &mut token,
                 );
                 return Some(id);
             }
-            selected -= adapters_dx11.len();
+            selected -= adapters.dx11.0.len();
         }
 
-        let _ = (selected, id_vulkan, id_metal, id_dx12, id_dx11);
+        let _ = selected;
         log::warn!("Some adapters are present, but enumerating them failed!");
         None
     }
 
+    pub fn enumerate_adapters(&self, inputs: AdapterInputs<F::Input>) -> Vec<AdapterId> {
+        let adapters = self.internal_enumerate_adapters(inputs);
+        let mut adapter_list = Vec::new();
+        let mut token = Token::root();
+
+        #[cfg(any(
+            not(any(target_os = "ios", target_os = "macos")),
+            feature = "gfx-backend-vulkan"
+        ))]
+        {
+            for raw in adapters.vulkan.0 {
+                let adapter = Adapter { raw };
+                log::info!("Adapter Vulkan {:?}", adapter.raw.info);
+                adapter_list.push(backend::Vulkan::hub(self).adapters.register_identity(
```

Comment:
I'm not sure if `register_identity` is really what we want to do here if we desire to just give a list of GPUs back to the user.

Reply:
if we are returning `AdapterInfo` structs, we can delay the `register_identity` to a point where we are actually assigning an ID to it

Reply:
Yes, I removed this in the latest commit.